File: src/client/js/agentguard/plugins/manager.js
"use strict";

const fs = require("fs");
const { CheckResult, BasePlugin } = require("./base");
const { getPluginClass, discoverPlugins } = require("./registry");
const { LLMInputPlugin } = require("./llm_before/llm_input");
const { LLMOutputPlugin } = require("./llm_after/llm_output");
const { ToolInvokePlugin } = require("./tool_before/tool_invoke");
const { ToolResultPlugin } = require("./tool_after/tool_result");

const PHASE_ORDER = ["llm_before", "llm_after", "tool_before", "tool_after", "global"];
const EVENT_PHASE = {
  llm_input: "llm_before",
  llm_output: "llm_after",
  tool_invoke: "tool_before",
  tool_result: "tool_after",
};
const BUILTIN_PLUGINS = {
  llm_input: LLMInputPlugin,
  llm_output: LLMOutputPlugin,
  tool_invoke: ToolInvokePlugin,
  tool_result: ToolResultPlugin,
};

function defaultPlugins() {
  return [];
}

function loadPluginConfig(source = null) {
  if (source == null) {
    return null;
  }
  let data;
  if (typeof source === "string") {
    data = JSON.parse(fs.readFileSync(source, "utf-8"));
  } else {
    data = { ...source };
  }
  const phases = data.phases;
  if (!phases || typeof phases !== "object" || Array.isArray(phases)) {
    throw new Error("plugin config must contain a 'phases' object");
  }
  const config = {};
  for (const phase of PHASE_ORDER) {
    if (phase in phases) {
      config[phase] = pluginSpecsForScope(phases[phase], "client");
    }
  }
  return config;
}

function pluginSpecsForScope(value, scope) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("plugin phase config must be an object with 'client' and 'server'");
  }
  if (!hasScope(value, "client") || !hasScope(value, "server")) {
    throw new Error("plugin phase config must include both 'client' and 'server'");
  }
  const specs = scopeValue(value, scope);
  if (specs == null) {
    return [];
  }
  if (!Array.isArray(specs)) {
    throw new Error(`plugin phase '${scope}' config must be a list`);
  }
  return [...specs];
}

function hasScope(value, scope) {
  return Object.prototype.hasOwnProperty.call(value, scope)
    || Object.prototype.hasOwnProperty.call(value, legacyScope(scope));
}

function scopeValue(value, scope) {
  if (Object.prototype.hasOwnProperty.call(value, scope)) {
    return value[scope];
  }
  return value[legacyScope(scope)];
}

function legacyScope(scope) {
  return scope === "client" ? "local" : "remote";
}

function buildPluginsByPhase(config = null) {
  if (!config) {
    return {};
  }
  const result = {};
  for (const [phase, specs] of Object.entries(config)) {
    result[phase] = specs.map(instantiatePlugin);
  }
  return result;
}

function instantiatePlugin(spec) {
  if (spec instanceof BasePlugin) {
    return spec;
  }
  if (typeof spec === "function") {
    return buildPlugin(spec);
  }
  if (typeof spec === "string") {
    discoverPlugins();
    const PluginClass = BUILTIN_PLUGINS[spec] || getPluginClass(spec);
    if (!PluginClass) {
      throw new Error(`invalid plugin config entry: ${String(spec)}`);
    }
    return buildPlugin(PluginClass);
  }
  if (spec && typeof spec === "object") {
    const target = spec.class || spec.plugin || spec.name;
    const kwargs = pluginKwargs(spec);
    const env = pluginEnv(spec);
    const PluginClass = typeof target === "function" ? target : BUILTIN_PLUGINS[target] || getPluginClass(target);
    if (!PluginClass) {
      throw new Error(`invalid plugin config entry: ${JSON.stringify(spec)}`);
    }
    return buildPlugin(PluginClass, { kwargs, env });
  }
  throw new Error(`invalid plugin config entry: ${String(spec)}`);
}

function pluginKwargs(spec) {
  const reserved = new Set(["class", "plugin", "name", "kwargs", "env"]);
  const kwargs = Object.fromEntries(Object.entries(spec).filter(([key]) => !reserved.has(key)));
  if (spec.kwargs != null && (typeof spec.kwargs !== "object" || Array.isArray(spec.kwargs))) {
    throw new Error(`plugin kwargs config must be an object: ${JSON.stringify(spec)}`);
  }
  return { ...kwargs, ...(spec.kwargs || {}) };
}

function pluginEnv(spec) {
  if (spec.env != null && (typeof spec.env !== "object" || Array.isArray(spec.env))) {
    throw new Error(`plugin env config must be an object: ${JSON.stringify(spec)}`);
  }
  return { ...(spec.env || {}) };
}

function buildPlugin(PluginClass, { kwargs = null, env = null } = {}) {
  const pluginKwargs = { ...(kwargs || {}) };
  const pluginEnv = { ...(env || {}) };
  try {
    return new PluginClass({ env: pluginEnv, ...pluginKwargs });
  } catch (_) {
    const plugin = new PluginClass();
    if (typeof plugin.bind_config === "function") {
      plugin.bind_config({ env: pluginEnv, ...pluginKwargs });
    }
    return plugin;
  }
}

class PluginManager {
  constructor({ plugins = null, config = null } = {}) {
    this.plugins_by_phase = plugins ? { global: [...plugins] } : buildPluginsByPhase(loadPluginConfig(config));
    this.refresh();
  }

  update_config(config = null) {
    this.plugins_by_phase = buildPluginsByPhase(loadPluginConfig(config));
    this.refresh();
  }

  updateConfig(config = null) {
    this.update_config(config);
  }

  add(plugin, phase = null) {
    const target = phase || inferPhase(plugin);
    this.plugins_by_phase[target] = this.plugins_by_phase[target] || [];
    this.plugins_by_phase[target].push(plugin);
    this.plugins.push(plugin);
  }

  refresh() {
    this.plugins = PHASE_ORDER.flatMap((phase) => this.plugins_by_phase[phase] || []);
  }

  run(event, context) {
    const phase = EVENT_PHASE[event.event_type] || "global";
    const phasePlugins = [...(this.plugins_by_phase[phase] || []), ...(this.plugins_by_phase.global || [])];
    const mergedSignals = [];
    let candidate = null;
    let isFinal = false;
    const metadata = {};
    for (const plugin of phasePlugins) {
      if (!plugin.applies(event)) {
        continue;
      }
      try {
        const result = plugin.check(event, context);
        for (const signal of result.risk_signals) {
          if (!mergedSignals.includes(signal)) {
            mergedSignals.push(signal);
          }
        }
        Object.assign(metadata, result.metadata || {});
        if (result.decision_candidate && (candidate === null || result.is_final)) {
          candidate = result.decision_candidate;
          isFinal = isFinal || result.is_final;
        }
      } catch (error) {
        metadata[`${plugin.name}_error`] = String(error.message || error);
      }
    }
    for (const signal of mergedSignals) {
      event.addSignal(signal);
    }
    return new CheckResult({
      decision_candidate: candidate,
      risk_signals: mergedSignals,
      is_final: isFinal,
      metadata,
    });
  }
}

function inferPhase(plugin) {
  for (const eventType of plugin.event_types || []) {
    const phase = EVENT_PHASE[eventType];
    if (phase) {
      return phase;
    }
  }
  return "global";
}

module.exports = {
  PHASE_ORDER,
  PluginManager,
  defaultPlugins,
  loadPluginConfig,
  load_plugin_config: loadPluginConfig,
};

File: src/client/python/agentguard/plugins/manager.py
"""Checker manager: run applicable checkers and merge results."""
from __future__ import annotations

import importlib
import inspect
import json
from pathlib import Path
from typing import Any

from agentguard.plugins.base import BaseChecker, CheckResult
from agentguard.plugins.registry import get_checker_class
from agentguard.schemas.context import RuntimeContext
from agentguard.schemas.events import EventType, RuntimeEvent

PHASE_ORDER = ("llm_before", "llm_after", "tool_before", "tool_after", "global")

_EVENT_PHASE = {
    EventType.LLM_INPUT: "llm_before",
    EventType.LLM_OUTPUT: "llm_after",
    EventType.TOOL_INVOKE: "tool_before",
    EventType.TOOL_RESULT: "tool_after",
}

def default_checkers() -> list[BaseChecker]:
    return []


def default_checker_config() -> dict[str, dict[str, list[Any]]]:
    return {}


def load_checker_config(source: str | Path | dict[str, Any] | None) -> dict[str, list[Any]]:
    if source is None:
        return {}
    elif isinstance(source, (str, Path)):
        path = Path(source)
        with path.open("r", encoding="utf-8") as fh:
            data = json.load(fh)
    else:
        data = dict(source)

    phases = data.get("phases")
    if not isinstance(phases, dict):
        raise ValueError("checker config must contain a 'phases' object")
    config: dict[str, list[Any]] = {}
    for phase in PHASE_ORDER:
        if phase in phases:
            config[phase] = _checker_specs_for_scope(phases.get(phase), "local")
    return config


def _checker_specs_for_scope(value: Any, scope: str) -> list[Any]:
    if not isinstance(value, dict):
        raise ValueError("checker phase config must be an object with 'local' and 'remote'")
    if "local" not in value or "remote" not in value:
        raise ValueError("checker phase config must include both 'local' and 'remote'")
    specs = value.get(scope)
    if specs is None:
        return []
    if not isinstance(specs, list):
        raise ValueError(f"checker phase '{scope}' config must be a list")
    return list(specs)


def build_checkers_by_phase(config: dict[str, list[Any]]) -> dict[str, list[BaseChecker]]:
    return {
        phase: [_instantiate_checker(spec) for spec in specs]
        for phase, specs in config.items()
    }


def _instantiate_checker(spec: Any) -> BaseChecker:
    if isinstance(spec, BaseChecker):
        return spec
    if isinstance(spec, type) and issubclass(spec, BaseChecker):
        return _build_checker(spec)
    if isinstance(spec, str):
        cls = get_checker_class(spec) or _load_checker_class(spec)
        return _build_checker(cls)
    if isinstance(spec, dict):
        target = spec.get("class") or spec.get("checker") or spec.get("name")
        kwargs = _checker_kwargs(spec)
        env = _checker_env(spec)
        if isinstance(target, str):
            cls = get_checker_class(target) or _load_checker_class(target)
        elif isinstance(target, type) and issubclass(target, BaseChecker):
            cls = target
        else:
            raise ValueError(f"invalid checker config entry: {spec!r}")
        return _build_checker(cls, kwargs=kwargs, env=env)
    raise ValueError(f"invalid checker config entry: {spec!r}")


def _checker_kwargs(spec: dict[str, Any]) -> dict[str, Any]:
    reserved = {"class", "checker", "name", "kwargs", "env"}
    kwargs = {key: value for key, value in spec.items() if key not in reserved}
    explicit_kwargs = spec.get("kwargs") or {}
    if not isinstance(explicit_kwargs, dict):
        raise ValueError(f"checker kwargs config must be an object: {spec!r}")
    kwargs.update(explicit_kwargs)
    return kwargs


def _checker_env(spec: dict[str, Any]) -> dict[str, Any]:
    env = spec.get("env") or {}
    if not isinstance(env, dict):
        raise ValueError(f"checker env config must be an object: {spec!r}")
    return dict(env)


def _build_checker(
    cls: type[BaseChecker],
    *,
    kwargs: dict[str, Any] | None = None,
    env: dict[str, Any] | None = None,
) -> BaseChecker:
    checker_kwargs = dict(kwargs or {})
    checker_env = dict(env or {})
    if _accepts_env_kwarg(cls):
        return cls(env=checker_env, **checker_kwargs)
    checker = cls(**checker_kwargs)
    checker.bind_config(env=checker_env, **checker_kwargs)
    return checker


def _accepts_env_kwarg(cls: type[BaseChecker]) -> bool:
    try:
        params = inspect.signature(cls.__init__).parameters.values()
    except (TypeError, ValueError):
        return False
    return any(param.kind == inspect.Parameter.VAR_KEYWORD for param in params) or any(
        param.name == "env" for param in params
    )


def _load_checker_class(path: str) -> type[BaseChecker]:
    module_name, _, class_name = path.rpartition(".")
    if not module_name or not class_name:
        raise ValueError(f"checker must be a builtin name or import path: {path}")
    module = importlib.import_module(module_name)
    cls = getattr(module, class_name)
    if not isinstance(cls, type) or not issubclass(cls, BaseChecker):
        raise TypeError(f"checker class must subclass BaseChecker: {path}")
    return cls


class CheckerManager:
    """Runs all applicable checkers and merges their CheckResults."""

    def __init__(
        self,
        checkers: list[BaseChecker] | None = None,
        *,
        config: str | Path | dict[str, Any] | None = None,
    ) -> None:
        if checkers is not None:
            self.checkers_by_phase = {"global": list(checkers)}
        else:
            self.checkers_by_phase = build_checkers_by_phase(load_checker_config(config))
        self._refresh_flat_checkers()

    def update_config(self, config: str | Path | dict[str, Any] | None) -> None:
        """Replace checker configuration for subsequent events."""
        self.checkers_by_phase = build_checkers_by_phase(load_checker_config(config))
        self._refresh_flat_checkers()

    def add(self, checker: BaseChecker, phase: str | None = None) -> None:
        target = phase or _infer_phase(checker)
        self.checkers_by_phase.setdefault(target, []).append(checker)
        self.checkers.append(checker)

    def _refresh_flat_checkers(self) -> None:
        self.checkers = [
            checker
            for phase in PHASE_ORDER
            for checker in self.checkers_by_phase.get(phase, [])
        ]

    def run(self, event: RuntimeEvent, context: RuntimeContext) -> CheckResult:
        merged_signals: list[str] = []
        candidate = None
        is_final = False
        meta: dict = {}
        phase = _EVENT_PHASE.get(event.event_type, "global")
        phase_checkers = list(self.checkers_by_phase.get(phase, []))
        phase_checkers.extend(self.checkers_by_phase.get("global", []))
        for checker in phase_checkers:
            if not checker.applies(event):
                continue
            try:
                res = checker.check(event, context)
            except Exception as exc:  # checkers must never break the flow
                meta[f"{checker.name}_error"] = str(exc)
                continue
            for s in res.risk_signals:
                if s not in merged_signals:
                    merged_signals.append(s)
            if res.metadata:
                meta.update(res.metadata)
            # Keep the strongest final candidate (first final wins).
            if res.decision_candidate and (candidate is None or res.is_final):
                candidate = res.decision_candidate
                is_final = is_final or res.is_final
        # Annotate the event with detected signals.
        for s in merged_signals:
            event.add_signal(s)
        return CheckResult(
            decision_candidate=candidate,
            risk_signals=merged_signals,
            is_final=is_final,
            metadata=meta,
        )


def _infer_phase(checker: BaseChecker) -> str:
    for event_type in checker.event_types:
        phase = _EVENT_PHASE.get(event_type)
        if phase:
            return phase
    return "global"
